tools/tslint-rules/lightweightTokensRule.ts

```diff
@@ -0,0 +1,198 @@
+    // consume them through an optionally lightweight token.
+    if (ctx.options[1].includes(symbol.getName())) {
+      continue;
+    }
+
+    // No lightweight tokens are available for external symbols. Framework might
+    // add tokens in the future, but now there is no way for most symbols.
+    if (symbol.valueDeclaration.getSourceFile().isDeclarationFile) {
+      continue;
+    }
+
+    // If the containing class injects itself, then lightweight tokens will not have
+    // any effect, and we skip the lint.
+    if (symbol.valueDeclaration === node.parent) {
+      continue;
+    }
+
+    // If the parameter has a `breaking-change` JSDoc that denotes the parameter
+    // to be made `required`, skip the lightweight token failure.
+    if (hasPendingRequiredJsdoc(param)) {
+      continue;
+    }
+
+    ctx.addFailureAtNode(param.type, failureMessage);
+  }
+}
+
+/** Gets an identifier that represents the given type node's resolved value. */
+function getInjectionType(type: ts.TypeNode): ts.Identifier|null {
+  // Unwraps union types with `null`. This is supported in Angular when
+  // `@Optional` is used. The `null` union type is ignored.
+  if (ts.isUnionTypeNode(type)) {
+    const nonNullTypes = type.types.filter(t => t.kind !== ts.SyntaxKind.NullKeyword);
+    return nonNullTypes.length === 1 ? getInjectionType(nonNullTypes[0]) : null;
+  }
+  if (!ts.isTypeReferenceNode(type)) {
+    return null;
+  }
+  const unwrapEntityName = (n: ts.EntityName): ts.Identifier => {
+    if (ts.isQualifiedName(n)) {
+      return unwrapEntityName(n.right);
+    }
+    return n;
+  };
+  return unwrapEntityName(type.typeName);
+}
+
+/**
+ * Analyzes the decorators of the given parameter. Returns whether it has
+ * an `@Inject` or `@Optional` decorator applied.
+ */
+function analyzeDecorators(param: ts.ParameterDeclaration):
+    {hasInject: boolean, isOptional: boolean} {
+  if (!param.decorators) {
+    return {hasInject: false, isOptional: false};
+  }
+  let hasInject = false;
+  let isOptional = false;
+  for (const decorator of param.decorators) {
+    const expr = decorator.expression;
+    if (!ts.isCallExpression(expr) || !ts.isIdentifier(expr.expression)) {
+      continue;
+    }
+    if (expr.expression.text === 'Inject') {
+      hasInject = true;
+    } else if (expr.expression.text === 'Optional') {
+      isOptional = true;
+    }
+  }
+  return {hasInject, isOptional};
+}
+
+/**
+ * Whether the given parameter is currently denoted with the `breaking-change`
+ * JSDoc tag and will be made required in a future version.
+ */
+function hasPendingRequiredJsdoc(param: ts.ParameterDeclaration): boolean {
+  const sourceFile = param.getSourceFile();
+  const comments = ts.getLeadingCommentRanges(sourceFile.text, param.pos)
+      ?.map(r => sourceFile.text.substring(r.pos, r.end));
+  return !!comments?.some(c => c.includes('@breaking-change') && c.includes('required'));
```

Comment:
@crisbeto I'd love to have your thoughts on this. It feels not very ideal, but it looks like we do not have a consistent pattern for detecting parameters that are soon becoming required in a future version.

Reply:
I'm not sure why we'd need this check. We can still decorate parameters with `Inject` without it being a breaking change.

Reply:
That's a good point. I was under the impression for parameters like below, we wouldn't want to bother about them using lightweight tokens as they are going to be required anyway soon (in which case the lightweight token is unnecessary then).

```ts
// @breaking-change 9.0.0 `_outsideClickDispatcher` parameter to be made required.
@Optional() private _outsideClickDispatcher?: OverlayOutsideClickDispatcher) { }
```

I don't have a strong opinion on this, but it feels unnecessary bothering about these cases when they are supposed to be required soon anyway.

Reply:
I don't think these cases are supposed to be marked as `Optional` to begin with. They have to be optional for TS so users that extend them don't get compilation errors, but they're still required for DI.

Reply:
Not sure I follow. These parameters obviously are _optional_ in runtime and at type-level. Also they are optional for Angular's DI using `@Optional`. Apparently this is just for the time being as in reality we'd like to make this required in the future, so it doesn't make sense to use a lightweight token for something that is required/retained anyway in the future.

Reply:
I meant that these shouldn't have `@Optional` in the first place. Their type needs to be optional so we don't introduce a breaking change, but in terms of DI they can still be required since we're the ones that provide them.